File: mirror/macos/ReMarkableMirror/WiFiRepairConfirmationPolicy.swift
enum WiFiRepairProbeResult: Equatable, Sendable {
    case success
    case authenticatedTabletPrerequisiteMismatch
    case identityRejected
    case other
}

enum WiFiRepairDisposition: Equatable, Sendable {
    case offline
    case repair
}

/// Prevents one transitional Wi-Fi capability mismatch from becoming a repair
/// requirement. Direct USB setup failures do not pass through this policy.
struct WiFiRepairConfirmationPolicy: Sendable {
    private static let requiredConsecutiveMatches = 2

    private var candidateRoute: VerifiedWiFiRoute?
    private var consecutiveMatches = 0

    mutating func record(
        route: VerifiedWiFiRoute?,
        result: WiFiRepairProbeResult
    ) -> WiFiRepairDisposition {
        guard let route,
              result == .authenticatedTabletPrerequisiteMismatch else {
            reset()
            return .offline
        }

        if candidateRoute != route {
            candidateRoute = route
            consecutiveMatches = 0
        }

        consecutiveMatches = min(
            consecutiveMatches + 1,
            Self.requiredConsecutiveMatches
        )
        return consecutiveMatches >= Self.requiredConsecutiveMatches
            ? .repair
            : .offline
    }

    mutating func reset() {
        candidateRoute = nil
        consecutiveMatches = 0
    }
}

File: mirror/windows/ReMarkableMirror/WifiRepairConfirmationPolicy.cs
namespace ReMarkableMirror;

/// <summary>
/// Prevents one transitional Wi-Fi capability result from being presented as
/// a durable tablet repair requirement. Direct USB setup failures do not pass
/// through this policy.
/// </summary>
internal sealed class WifiRepairConfirmationPolicy
{
    private const int RequiredConsecutiveMatches = 2;
    private int _consecutiveMatches;

    public bool Record(bool identityAuthenticated, bool tabletPrerequisiteMismatch)
    {
        if (!identityAuthenticated || !tabletPrerequisiteMismatch)
        {
            Reset();
            return false;
        }

        _consecutiveMatches = Math.Min(
            _consecutiveMatches + 1,
            RequiredConsecutiveMatches);
        return _consecutiveMatches >= RequiredConsecutiveMatches;
    }

    public void Reset() => _consecutiveMatches = 0;
}
